design-proposals/configurable-features.md

```diff
@@ -0,0 +1,233 @@
+accomplished while achieving [eventually consistency](https://en.wikipedia.org/wiki/Eventual_consistency). This forces
+us to avoid the feature state control checking on webhooks and moving the feature state control closer to the
+responsible code. Moreover, it has to be decided how the system should behave if a VMI is
+requiring a feature in a state different from what was configured in the KubeVirt CR, or what should happen if the
+configuration of a feature in use is changed. (see matrix below).
+
+## Goals
+
+- Get a clear understanding about the features configuration status.
+- Establish how the feature configurables should work.
+- Describe how the system should react in these scenarios in the case that the VMI exposes an API field to configure the
+  feature status:
+    - A feature in KubeVirt is set to state A and a VMI requests the feature to be in state B.
+    - A feature in KubeVirt is set to state A, there are running VMIs using the feature in state A, and the feature is 
+      changed in KubeVirt to state B.
+    - A feature in KubeVirt is set to state A, and pending VMIs want to use it.
+    - A feature in KubeVirt is set to state A, and running VMIs using the feature in state B wants to live migrate.
+- Graduate features status swapping from features gates to configurables.
+
+## Non Goals
+
+- Describe how features protected with features gates should work.
+- Change how feature gates are managed. Feature gating and configuration are two completely distinct issues.
+
+## Definition of Users
+
+Development contributors.
+
+Cluster administrators.
+
+## User Stories
+
+* As a cluster administrator, I want to be able to change the cluster-wide state of a feature by editing configurables.
+
+* As VMI owner, I want to use a given feature.
+
+* As a VMI owner / cluster admin, I want to understand what's the current configuration of the various features.
+
+## Repos
+
+Kubevirt/Kubevirt
+
+# Design
+
+The feature status swapping must be done by adding new fields to the KubeVirt CR under
+`spec`:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+[...]
+spec:
+  certificateRotateStrategy: {}
+  feature-A:  {}
+  feature-C:
+    configA: integer
+    configB: string
+[...]
+```
+Please note that if the feature spec field is not present, the feature status is assumed to be completely disabled.
+For instance, in the KubeVirt CR manifest provided above, `feature-B` is not enabled.
+
+The VMI object may or may not include a configuration field inside the relevant spec.
+
+> **NOTE:** The inclusion of these new KubeVirt API fields should be carefully considered and justified. The feature
+> configurables should be avoided as much as possible.
+
+Moreover, the KubeVir CR `status` field should clearly indicate the current state of each feature, providing a
+comprehensive overview of the operational status of these features:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+[...]
+status:
```

Comment:
this is too abstract for me. How would this look like for existing features?

Reply:
For instance, let's suppose that we have `downwardMetrics` and `maxAllowedCPUsPerVM` feature. The `maxAllowedCPUsPerVM` controls the maximum amount of CPUs that a given VM can get. We enable the `downwardMetrics` and we want to restrict that any given VM just can get 2 CPUs. The Kubevirt CR will looks like:

```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
[...]
spec:
  downwardMetrics: {} # this means "enabled"
  maxCPUsPerVM: 2
[...]
status:
  featureStatus:
    downwardMetrics:
      status: Enabled
    maxCPUsPerVM:
      status: 2 CPUs
```

WDYT?

Reply:
No, it shouldn't differ. This idea was proposed here: https://github.com/kubevirt/community/pull/316#discussion_r1727443301

Reply:
All right. I will drop it.

Reply:
Dropped.